content/blog/taksim-modes-makam.md

```diff
@@ -0,0 +1,664 @@
+we'll keep that as-is for now and revisit that later.
+
+Transcribing the rest of the rules, we proceed similarly for the
+checking rules, and there is not much we need to change for synthesis rules.
+
+```makam
+type_synth X A :-
+  type_of_var X A.
+
+type_check E B :-
+  type_synth E A, unify A B.
+
+type_synth (annot E A) A :-
+  type_check E A.
+
+type_check eunit T :-
+  pattern_match tunit T.
+
+type_synth (app E1 E2) B :-
+  type_synth E1 (arrow A B),
+  type_check E1 A.
+```
+
+Let's try a few examples to see what we did:
+
+```makam
+type_check (lambda (fun x => x)) (arrow tunit tunit) ?
+type_check (lambda (fun x => x)) T ?
+type_check (lambda (fun x => x)) (arrow tunit X) ?
+type_check (lambda (fun x => x)) (arrow tunit (arrow tunit tunit)) ?
+```
+
+These all behave as expected.
+
+```makam-hidden
+>> type_check (lambda (fun x => x)) (arrow tunit tunit) ?
+>> Yes.
+
+>> type_check (lambda (fun x => x)) T ?
+>> Impossible.
+
+>> type_check (lambda (fun x => x)) (arrow tunit X) ?
+>> Yes:
+>> X := tunit.
+
+>> type_check (lambda (fun x => x)) (arrow (arrow tunit tunit) X) ?
+>> Yes:
+>> X := arrow tunit tunit.
+
+>> type_check (lambda (fun x => x)) (arrow tunit (arrow tunit tunit)) ?
+>> Impossible.
+```
+
+```makam-hidden
+%end.
+```
+
+So the recipe for enforcing a proper mode is pretty simple: for input arguments, we should use pattern matching rather than unification.
+Is there some way that we can extend Makam to support this recipe?
+
+## modal makam
+
+```makam-hidden
+(* stdlib extensions required for this. will open a PR for Makam soon *)
+
+%extend refl.
+
+decompose_term : [Full Head] Full -> Head -> args Head Full -> prop.
+decompose_term Term Head Args :-
+  refl.headargs Term Head ArgsDyn,
+  dyn.to_args ArgsDyn Args.
+
+recompose_term : [Full Head] Head -> args Head Full -> Full -> prop.
+recompose_term Head Args Term :-
+  args.applyfull Head Args Term.
+
+%end.
+```
+
+```makam-hidden
+%extend modes_v1.
+```
+
+Here's a first approach: we can do some meta-programming in Makam, transforming predicates that utilize a specific mode according to the recipe. We will have to perform that transformation for all rules of the relevant predicates.
+
+Let's see how this would work. First of all, we can define `mode` as a predicate, that will be used to specify the mode for predicates. We'll make light use of GADTs to make sure that when specifying a mode, the number of arguments we specify matches the number of arguments of the predicate.
+
+```makam
+argument_mode : type.
+input, output : argument_mode.
+
+mode_spec : type -> type.
+nil : mode_spec prop.
+cons : argument_mode -> mode_spec B -> mode_spec (A -> B).
+
+mode : [A] A -> mode_spec A -> prop.
+```
+
+Now we should be able to specify modes for our predicates:
+
+```makam
+type_check : expr -> typ -> prop.
+mode type_check [input, input].
+
+type_synth : expr -> typ -> prop.
+mode type_synth [input, output].
+```
```

Comment:
This is really fancy and cool! Haha. How is `mode` defined though? Very curious! 🤯

Reply:
:+1: Yeah this trick needs a bit more explanation. `mode` serves more as a way to capture information about predicates rather than an executable predicate in itself.